biblib/views/document_view.py

```diff
@@ -44,20 +45,92 @@ def add_document_to_library(cls, library_id, document_data):
 
             start_length = len(library.bibcode)
 
-            library.add_bibcodes(document_data['bibcode'])
-
-            session.add(library)
-            session.commit()
-
-            current_app.logger.info('Added: {0} is now {1}'.format(
-                document_data['bibcode'],
-                library.bibcode)
-            )
-
+            #Validate supplied bibcodes to confirm they exist in SOLR
+            valid_bibcodes = []
+            page_size = min(current_app.config.get('BIGQUERY_MAX_ROWS', len(document_data['bibcode'])), len(document_data['bibcode']))
+            #Check if there are more than the allowed number of bibcodes. Iterate over pages if needed.
+            pages = len(document_data['bibcode']) // page_size + (len(document_data['bibcode']) % page_size > 0)
+            
+            for page in range(0, pages):
+                solr_resp, status_code = cls.query_valid_bibcodes(document_data['bibcode'], start=page*page_size, rows=min(page_size, len(document_data['bibcode'])))
+
+                if "error" in solr_resp.keys():
+                    #If SOLR request fails, pass the error back to the user
+                    current_app.logger.error("Failed to retrieve bibcodes with error: {}".format(solr_resp.get("error")))
+                    output_dict = {"error": solr_resp.get("error"), "number_added": 0, "status": status_code}
+                    valid_bibcodes += []
+                    return output_dict
+                else:
+                    #If SOLR query succeeds generate list of valid bibcodes from response
+                    add_bibcodes = [doc.get('bibcode') for doc in solr_resp.get('docs', {})]
+                    if add_bibcodes:
+                        valid_bibcodes += add_bibcodes
+                        current_app.logger.debug("Found the following valid bibcodes: {}".format(add_bibcodes))
+                    #Added additional check to prevent unnecessary calls to bigquery.
+                    if len(add_bibcodes) < current_app.config.get('BIGQUERY_MAX_ROWS'):
```

Comment:
It's mainly there as a check for if a user specifies a large number of invalid bibcodes. This will stop paging early so we don't have unnecessary calls to SOLR. I don't expect it to be used often but thought we needed something to prevent someone sending an incredibly large file of junk and just wasting cycles.